src/precompiles/modexp.rs

```diff
@@ -3,52 +3,80 @@ use crate::precompiles::{
-                    exp32_bytes.push(0u8);
-                }
-            }
-        }
-        let exp32 = U256::from(exp32_bytes.as_slice());
+impl<HF: HardFork> ModExp<HF> {
+    fn calc_iter_count(exp_len: U256, base_len: U256, bytes: &[u8]) -> U256 {
+        let start = base_len.as_usize() + 96;
+        let end = core::cmp::min(32, exp_len.as_usize()) + start;
+        let exp = U256::from(&bytes[start..end]);
 
-        if exp_len <= U256::from(32) && exp32 == U256::zero() {
+        if exp_len <= U256::from(32) && exp == U256::zero() {
             U256::zero()
         } else if exp_len <= U256::from(32) {
-            U256::from(exp32.bits())
+            U256::from(exp.bits()) - U256::from(1)
         } else {
             // else > 32
-            U256::from(8) * (exp_len - U256::from(32)) + U256::from(exp32.bits())
+            U256::from(8) * (exp_len - U256::from(32)) + U256::from(exp.bits()) - U256::from(1)
         }
     }
 
-    fn mult_complexity(x: U256) -> Result<U256, ExitError> {
+    fn run_inner(input: &[u8]) -> Result<Vec<u8>, ExitError> {
+        let base_len = U256::from(&input[0..32]).as_usize();
+        let exp_len = U256::from(&input[32..64]).as_usize();
+        let mod_len = U256::from(&input[64..96]).as_usize();
+
+        let base_start = 96;
+        let base_end = base_len + base_start;
+        let base_bytes = &input[base_start..base_end];
+
+        let exp_start = base_len + 96;
+        let exp_end = exp_len + exp_start;
+        let exp_bytes = &input[exp_start..exp_end];
+
+        let mod_start = base_len + exp_len + 96;
+        let mod_end = mod_len + mod_start;
+        let mod_bytes = &input[mod_start..mod_end];
+
+        let base = BigUint::from_bytes_be(&base_bytes);
+        let exponent = BigUint::from_bytes_be(&exp_bytes);
+        let modulus = BigUint::from_bytes_be(&mod_bytes);
+
+        let output = {
+            let computed_result = base.modpow(&exponent, &modulus).to_bytes_be();
+            // The result must be the same length as the input modulus.
+            // To ensure this we pad on the left with zeros.
+            if mod_len > computed_result.len() {
+                let diff = mod_len - computed_result.len();
+                let mut padded_result = Vec::with_capacity(mod_len);
+                padded_result.extend(core::iter::repeat(0).take(diff));
+                padded_result.extend_from_slice(&computed_result);
+                padded_result
+            } else {
+                computed_result
+            }
+        };
+
+        Ok(output)
+    }
+}
+
+impl ModExp<Byzantium> {
+    fn mul_complexity(x: U256) -> Result<U256, ExitError> {
         if x <= U256::from(64) {
             Ok(x * x)
         } else if x <= U256::from(1_024) {
             Ok(x * x / U256::from(4) + U256::from(96) * x - U256::from(3_072))
         } else {
-            let (sqroot, overflow) = x.overflowing_mul(x);
+            let (x_sq, overflow) = x.overflowing_mul(x);
```

Comment:
Overflow check is no longer necessary because the max value of `x` is `u32::MAX` and `(2^32)^2 = 2^64 < 2^256`.